deeppavlov/pipeline_manager/observer.py

```diff
@@ -0,0 +1,180 @@
+            name: name of the experiments.
+            root: path to root folder.
+            info: additional information that you want to add to the log, the content of the dictionary
+             does not affect the algorithm
+            date: date of the experiment.
+    """
+
+    def __init__(self, name: str, root: Union[str, Path], info: dict, date: str, plot: bool) -> None:
+        """ Initializes the log, creates a folders tree and files necessary for the observer to work. """
+
+        self.exp_name = name
+        self.exp_inf = info
+        self.root = root
+        self.date = date
+        self.plot = plot
+
+        # tmp parameters
+        self.pipe_ind = 0
+        self.pipe_conf = None
+        self.model = None
+        self.pipe_res = None
+        self.pipe_time = None
+        self.batch_size = None
+
+        # build folder dependencies
+        self.save_path = self.root / self.date / self.exp_name / 'checkpoints'
+        self.log_path = self.root / date / self.exp_name
+        self.exp_file = self.log_path / (self.exp_name + '.json')
+        self.log_file = self.log_path / 'logs.jsonl'
+
+        if not self.save_path.is_dir():
+            self.save_path.mkdir(parents=True)
+        if self.plot:
+            if not (self.log_path / 'images').is_dir():
+                (self.log_path / 'images').mkdir()
+
+        self.exp_info = OrderedDict(date=date,
+                                    exp_name=self.exp_name,
+                                    root=str(self.root),
+                                    info=self.exp_inf,
+                                    number_of_pipes=None,
+                                    metrics=None,
+                                    target_metric=None,
+                                    dataset_name=None)
+        self.log = None
+
+    def save_exp_info(self):
+        """ Write exp_info in json file. """
+        with open(str(self.exp_file), 'w') as exp_file:
+            json.dump(self.exp_info, exp_file)
+
+    def exp_time(self, time: str) -> None:
+        """
+        Adding the time duration of the experiment in log file.
+
+        Args:
+            time: the time duration of the experiment
+
+        Returns:
+            None
+        """
+        with open(str(self.exp_file), 'r') as old_log:
+            old_log = json.load(old_log)
+
+        old_log['full_time'] = time
+        with open(str(self.exp_file), 'w') as log_file:
+            json.dump(old_log, log_file)
+
+    def tmp_reset(self) -> None:
+        """ Reinitialize temporary attributes. """
+        # tmp parameters
+        self.pipe_ind = 0
+        self.pipe_conf = None
+        self.model = None
+        self.pipe_res = None
+        self.pipe_time = None
+        self.batch_size = None
+        self.log = None
+
+    def write(self) -> None:
+        """ Write pipeline logs in jsonl file. """
+        if self.log_file.is_file():
+            with open(str(self.log_file), 'a') as logs:
```

Comment:
This whole method can be rewritten as something like:
```python3
with self.log_file.open('a', encoding='utf8') as f:
    print(json.dumps(self.log), file=f)
```

Reply:
Thanks, i am stupid.